app/src/main/kotlin/be/zvz/kookie/inventory/transaction/InventoryTransaction.kt

```diff
@@ -0,0 +1,217 @@
+
+import be.zvz.kookie.inventory.Inventory
+import be.zvz.kookie.inventory.transaction.action.InventoryAction
+import be.zvz.kookie.inventory.transaction.action.SlotChangeAction
+import be.zvz.kookie.item.Item
+import be.zvz.kookie.player.Player
+import com.koloboke.collect.map.hash.HashObjIntMaps
+import com.koloboke.collect.map.hash.HashObjObjMaps
+import kotlin.math.min
+
+class InventoryTransaction(val source: Player, actions: MutableList<InventoryAction>) {
+    var actions: MutableList<InventoryAction> = mutableListOf()
+        private set
+
+    var inventories: MutableList<Inventory> = mutableListOf()
+    var hasExecuted: Boolean = false
+        private set
+
+    init {
+        actions.forEach {
+            addAction(it)
+        }
+    }
+
+    fun addAction(action: InventoryAction) {
+        if (!actions.contains(action)) {
+            actions.add(action)
+            action.onAddToTransaction(this)
+        } else {
+            throw IllegalArgumentException("Tried to add the same action to a transaction twice")
+        }
+    }
+
+    private fun shuffleActions() {
+        actions.shuffle()
+    }
+
+    fun addInventory(inventory: Inventory) {
+        if (!inventories.contains(inventory)) {
+            inventories.add(inventory)
+        }
+    }
+
+    fun matchItems(needItems: MutableList<Item>, haveItems: MutableList<Item>) {
+        actions.forEach {
+            if (!it.targetItem.isNull()) {
+                needItems.add(it.targetItem)
+            }
+            try {
+                it.validate(source)
+            } catch (e: TransactionValidationException) {
+                throw TransactionValidationException("${it::class.java.simpleName}: " + e.message)
+            }
+            if (!it.sourceItem.isNull()) {
+                haveItems.add(it.sourceItem)
+            }
+        }
+        needItems.apply {
+            val iterator1 = iterator()
+            while (iterator1.hasNext()) {
+                val needItem = iterator1.next()
+                haveItems.apply {
+                    val iterator2 = iterator()
+                    while (iterator2.hasNext()) {
+                        val haveItem = iterator2.next()
+                        if (needItem.equals(haveItem)) {
+                            val amount = min(needItem.count, haveItem.count)
+                            needItem.count = needItem.count - amount
+                            haveItem.count = haveItem.count - amount
+                            if (haveItem.count == 0) {
+                                iterator1.remove()
+                            }
+                            if (needItem.count == 0) {
+                                iterator1.remove()
+                                return@apply
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    protected fun squashDuplicateSlotChanges() {
```

Comment:
Member with protected visibility in final class is private. Consider using private or internal as a modifier.